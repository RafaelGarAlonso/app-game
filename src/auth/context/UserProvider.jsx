import { UserContext } from './UserContext';
import { INIT_USER, INIT_USERS_TABLE } from '../../constants/users'

export const UserProvider = ({ children }) => {

  const getUser = () => {
    const localStorageData = localStorage.getItem('currentUser');
    return localStorageData ? JSON.parse(localStorageData) : INIT_USER;
  };

  const setNewUserSession = ({ userName }) => {
    const listOfUsers = getListOfUsers();

    for (const [index, user] of listOfUsers.entries()) {
      if (user.userName === userName) {
        const { userName, points } = listOfUsers[index];
        localStorage.setItem('currentUser', JSON.stringify({ userName, points }));
        return false;
      }
    };

    localStorage.setItem('currentUser', JSON.stringify({ userName, points: 0 }));
    return true;
  };

  const getAppLoaded = () => {
    return sessionStorage.getItem('appLoaded');
  };

  const setAppLoaded = (loaded) => {
    sessionStorage.setItem('appLoaded', loaded);
  };

  const userEarnPoints = () => {
    const user = getUser();
    localStorage.setItem('currentUser', JSON.stringify({ userName: user.userName, points: user.points + 1 }));
  };

  const getListOfUsers = () => {
    const listOfUsers = localStorage.getItem('listOfUsers');
    return listOfUsers ? JSON.parse(listOfUsers) : null;
  };

  const setNewListOfUsers = (listOfUsers = INIT_USERS_TABLE) => {
    localStorage.setItem('listOfUsers', JSON.stringify(listOfUsers));
  };

  const setPointsToListUsers = (currentUser) => {
    const listOfUsers = getListOfUsers();

    for (let [index, user] of listOfUsers.entries()) {
      if (user.userName === currentUser.userName) {
        listOfUsers[index].points = currentUser.points;
        break;
      }
    };

    setNewListOfUsers(listOfUsers);
  };

  return (
    <UserContext.Provider value={{
      setNewUserSession,
      getUser,
      setAppLoaded,
      getAppLoaded,
      userEarnPoints,
      getListOfUsers,
      setNewListOfUsers,
      setPointsToListUsers
    }}>
      {children}
    </UserContext.Provider>
  )
}