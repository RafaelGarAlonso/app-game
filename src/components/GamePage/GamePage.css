.navbar-info {
    display: flex;
}

.button-back {
    position: absolute;
    width: 100px;
    margin: 20px;
    bottom: 0;
    right: 0;
}

.game-buttons {
    width: 100%;
    position: absolute;
    top: 50%;
    left: 50%;
    text-align: center;
    transform: translate(-50%, -50%);
}

.turn {
    font-size: 30px;
    font-weight: bold;
    margin-bottom: 20px;
    text-decoration: underline;
}

.game-images {
    border: 1px dotted #747474;
    padding: 5px;
    width: 95px;
    height: 95px;
    margin: 5px;
    object-fit: contain;
}

.your-chosen {
    color: brown;
    margin-bottom: 10px;
    font-weight: 500;
}

.bot-chosen {
    color: blue;
    margin-bottom: 50px;
    font-weight: 500;
}

.play-again {
    margin-top: 50px;
    width: 150px;
}

.result {
    width: 300px;
    color: #fff;
    margin: 0 auto;
    padding: 15px;
    border-radius: 20px;
}

.result-win {
    background-color: green;
}

.result-lose {
    background-color: red;
}

.result-tie {
    color: #000;
    background-color: yellow;
}

.rock {
    background-color: red;
}

.button-rock {
    margin-right: 25px;
}

.button-scissors {
    margin-top: 20px;
}

.paper {
    background-color: yellow;
}

.scissors {
    background-color: green;
}

.progress-bot {
    width: 80%;
    margin: 0 auto;
}

button {
    margin: 5px;
    padding: 0;
}