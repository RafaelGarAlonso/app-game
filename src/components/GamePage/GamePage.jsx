import { Container, Button, Image, Row, Col } from 'react-bootstrap';
import { useContext } from 'react';
import { Navbar } from '../Navbar/Navbar';
import { AnimatedProgressBar } from '../ProgressBar/ProgressBar';
import { useNavigate } from 'react-router-dom';
import { UserContext } from '../../auth/context/UserContext';
import { useState } from 'react';
import { listOfPlays } from '../../constants/users';
import './GamePage.css';

export const GamePage = () => {
    const { getUser, userEarnPoints, setPointsToListUsers } = useContext(UserContext);
    const { userName, points } = getUser();
    const [turn, setTurn] = useState("you");
    const [yourSelection, setYourSelection] = useState("");
    const [botSelection, setBotSelection] = useState("");
    const [resultOfGame, setResultOfGame] = useState("");
    const navigate = useNavigate();

    const onBack = () => {
        navigate('/home');
    }

    const playAgain = () => {
        setTurn("you");
    }

    const play = (hand) => {
        setYourSelection(hand);
        setTurn("bot");

        const resultOfBot = listOfPlays[Math.floor((Math.random() * listOfPlays.length))];
        setBotSelection(resultOfBot);

        if (hand === resultOfBot) return setResultOfGame("TIE!");
        if (hand === "rock" && resultOfBot === "scissors" ||
            hand === "paper" && resultOfBot === "rock" ||
            hand === "scissors" && resultOfBot === "paper") {
            return setResultOfGame("YOU WIN!");
        } else {
            setResultOfGame("YOU LOSE!");
        }
    };

    const onResults = () => {
        setTimeout(() => {
            setTurn("result");
            if (transformResultOfGame() === "win") {
                userEarnPoints();
                setPointsToListUsers(getUser());
            }

        }, 1000);
    };

    const turnTemplate = (turn) => {
        switch (turn) {
            case "you":
                return yourTurnTemplate();
            case "bot":
                return turnBotTemplate();
            case "result":
                return resultOfGameTemplate();
        }
    };

    const yourTurnTemplate = () => {
        return (
            <>
                <Row>
                    <Col>
                        <button className="button-rock" onClick={() => play("rock")}><Image className="game-images rock" thumbnail src="assets/rock.png" /></button>
                        <button className="button-paper" onClick={() => play("paper")}><Image className="game-images paper" thumbnail src="assets/paper.png" /></button>
                    </Col>
                </Row>
                <Row>
                    <Col>
                        <button className="button-scissors" onClick={() => play("scissors")}><Image className="game-images scissors" thumbnail src="assets/scissors.png" /></button>
                    </Col>
                </Row>
            </>
        )
    };

    const turnBotTemplate = () => {
        return (
            <div className="progress-bot">
                <AnimatedProgressBar onComplete={onResults} timeToComplete={10} />
            </div>
        )
    };

    const resultOfGameTemplate = () => {
        return (
            <>
                <h3 className="your-chosen">You has chosen: {yourSelection}</h3>
                <h3 className="bot-chosen">Bot has chosen: {botSelection}</h3>
                <h1 className={`result result-${transformResultOfGame()}`}>{resultOfGame}</h1>
            </>
        )
    };

    const turnText = (turn) => {
        if (turn === "you") return "Your turn!";
        if (turn === "bot") return "Bot turn!";
        if (turn === "result") return "Result of game:";
        return "";
    };

    const transformResultOfGame = () => {
        return resultOfGame.slice(-5).trim().replace("!", '').toLocaleLowerCase();
    };

    return (
        <>
            <Navbar userName={userName} points={points} />
            <Container className="container-game" fluid>
                <div className="game-buttons">
                    <div className="turn">
                        {turnText(turn)}
                    </div>
                    {turnTemplate(turn)}
                    {
                        turn === "result" ?
                            <Button
                                onClick={playAgain}
                                className="play-again"
                                size="lg"
                                variant="danger">
                                Play again
                            </Button>
                            : <></>
                    }
                </div>
            </Container>
            <Button onClick={onBack} className="button-back" size="lg" variant="dark">Exit</Button>
        </>
    )
}