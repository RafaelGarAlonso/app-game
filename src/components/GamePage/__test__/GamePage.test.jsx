import { render, fireEvent, waitFor } from '@testing-library/react';
import { GamePage } from '../GamePage';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from '../../../auth/context/UserProvider';

const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...(jest.requireActual('react-router-dom')),
    useNavigate: () => mockedUsedNavigate,
}));

describe('Test <GamePage />', () => {

    beforeEach(() => {
        jest.useFakeTimers();
        jest.spyOn(global, 'setInterval');
        jest.spyOn(global, 'clearInterval');
    });

    const renderTemplate = () => {
        return render(
            <UserProvider>
                <MemoryRouter>
                    <GamePage />
                </MemoryRouter>
            </UserProvider>
        );
    }

    test('Check snapshot component page', () => {
        const { container } = renderTemplate();
        expect(container).toMatchSnapshot();
    });

    test('It should click on button back and navigate to /home', () => {
        const { container } = renderTemplate();
        const backButton = container.querySelector(".button-back");

        fireEvent.click(backButton);
        expect(mockedUsedNavigate).toBeCalled();
    });

    test('It should click on rock button option and load bot progress-bar', async () => {
        const { container } = renderTemplate();
        const rockButton = container.querySelector(".button-rock");

        fireEvent.click(rockButton);

        await waitFor(() => {
            const progressBar = container.querySelector(".progress-bot");
            expect(progressBar.innerHTML).not.toBe("");
        });
    });

    test('It should click on scissors button option and load bot progress-bar', async () => {
        const { container } = renderTemplate();
        const rockButton = container.querySelector(".button-scissors");

        fireEvent.click(rockButton);

        await waitFor(() => {
            const progressBar = container.querySelector(".progress-bot");
            expect(progressBar.innerHTML).not.toBe("");
        });
    });

    test('It should click on paper button option and load bot progress-bar', async () => {
        const { container } = renderTemplate();
        const rockButton = container.querySelector(".button-paper");

        fireEvent.click(rockButton);

        await waitFor(() => {
            const progressBar = container.querySelector(".progress-bot");
            expect(progressBar.innerHTML).not.toBe("");
        });
    });

    // TODO: Render result case when callback complete is executed
    // test('It should click on rock button option and show result of game', async () => {
    //     const { container } = renderTemplate();
    //     const rockButton = container.querySelector(".button-rock");
    //     const delay = 1500;
        
    //     fireEvent.click(rockButton);
        
    //     act(() => {
    //         jest.advanceTimersByTime(delay - 500);
    //     });
        
    //     await waitFor(() => {
    //         const progressBot = container.querySelector(".progress-bot");
    //         console.log(progressBot.props)
    //         screen.debug()
    //     });
    // });
});