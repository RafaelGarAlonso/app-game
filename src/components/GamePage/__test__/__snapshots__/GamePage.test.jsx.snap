// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <GamePage /> Check snapshot component page 1`] = `
<div>
  <div
    class="navbar"
  >
    <div
      class="name"
    >
      Name: 
    </div>
    <div
      class="points"
    >
      Points: 
      0
    </div>
  </div>
  <div
    class="container-game container-fluid"
  >
    <div
      class="game-buttons"
    >
      <div
        class="turn"
      >
        Your turn!
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <button
            class="button-rock"
          >
            <img
              class="game-images rock img-thumbnail"
              src="assets/rock.png"
            />
          </button>
          <button
            class="button-paper"
          >
            <img
              class="game-images paper img-thumbnail"
              src="assets/paper.png"
            />
          </button>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <button
            class="button-scissors"
          >
            <img
              class="game-images scissors img-thumbnail"
              src="assets/scissors.png"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
  <button
    class="button-back btn btn-dark btn-lg"
    type="button"
  >
    Exit
  </button>
</div>
`;
